docs/TensorDenotation.md

```diff
@@ -0,0 +1,48 @@
+# Tensor Denotation
+
+Tensor Denotation is an experimental attempt along with [Dimension Denotation](DimensionDenotation.md) to give tensor semantic description and enable model authors to describe what the input and output tensors are.
+
+## Motivation
+
+The motivation of such a mechanism can be illustrated via a simple example. In the the neural network SqueezeNet, it takes in an NCHW image input float[1,2,244,244] and produces a output float[1,1000,1,1]:
+
+```
+input_in_NCHW -> data_0 -> SqueezeNet() -> output_softmaxout_1
+```
+
+In order to run this model the user needs a lot of information.    In this case the user needs to know:
+* the input is an image
+* the image is in the format of NCHW
+* the color channels are in the order of bgr
+* the pixel data is 8 bit
+* the pixel data is normalized as values 0-255
+
+This proposal consists of three key components to provide all of this information: Tensor Denotation, [Dimension Denotation](DimensionDenotation.md), and [model metadata](MetadataProps.md).  Each of which will be discussed in detail.
+
+## Tensor Denotatio Definition
+
+To begin with, we define a set of semantic types that define what models generally consume as inputs and produce as outputs.
+
+Specifically, in our first proposal we define the following set of standard denotations:
+
+1. `IMAGE` describes that a tensor holds an image.  You can use dimension denotation to learn more about the layout of the image, and also the optional model metadata_props.
+2. `AUDIO` describes that a tensor holds an audio clip.   
+3. `TEXT` describes that a tensor holds a block of text.
+
+Model authors SHOULD add tensor denotation to inputs and outputs for the model.
+
+## Denotation Propagation
+
+For tensors, Denotation Propagation does not automatically occur.   It is used to describe the initial input or final outputs, but as data flows through the graph no inference is made to propogate if the data still holds an image (for example).   A model builder or conversion tool MAY apply propagation manually in the model if they knows that subsequent tensors share the same semantic denotation.
+
+## Denotation Verification
+
+Denotation Verification is not enforced.   It is simply a method for model authors to indicate to model consumers what they should be passing in and what they should be expecting out.  No error is reported if you do not actually pass in an image (for example).
+
+## Dimension Denotation
+
+Tensor denotation can be combined with the new experimental feature of Dimension Denotation.  For example if the Tensor Denotation is `IMAGE`, then that tensor SHOULD also have [Dimension Denotation](DimensionDenotation.md) stating the NCHW channel layout.  (described as [`DATA_BATCH`, `DATA_CHANNEL`, `DATA_FEATURE`, `DATA_FEATURE`]).
```

Comment:
NCHW format on a model level is not easily enforceable and thus brings about nightmares in terms of compliance checking. ONNX, as it stands today, only enforces node-level input format, which does not translate to any restrictions on model level format. C.f. my comment copied from the thread discussing Dimension Denotation:


1. > ONNX as a standard will keep as it is to support only NCHW.

    We need to be careful about what you mean. There are no operations supporting NHWC format in ONNX; however, that does not preclude models from supporting NHWC format input. In fact, you cannot preclude such use case as long as you have a **transpose** in the operator set because you can always transpose the input once and enter the NCHW world thereafter. Thus, unless you explicitly pattern match and prohibit these cases in the checker, there is no way to prevent a user from creating  ONNX **models** that accept NHWC input but uses NCHW **operators**. Thus it is my opinion that 
    > This means all ONNX models should be in NCHW format,

    does not hold true as an implication from your previous statement.

Reply:
agreed.   with this we are not proposing enforcing or being strict about NCHW, but instead allowing enough metadata and denotation for a model builder to let a model consumer know what to do.    This PR is less about NCHW (dimension denotation talks about that).  This PR is more about the other things metadata talks about (bitmap pixel format, etc..etc..) .

thoughts?   Would you propose changing anything in the docs or strings for this PR ?  thx !